File: 0496-next-greater-element-i/0496-next-greater-element-i.java
import java.util.HashMap;
import java.util.Stack;

class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int n = nums1.length;
        int m = nums2.length;
        int arr[] = new int[n];
        HashMap<Integer, Integer> hm = new HashMap<>();
        Stack<Integer> st = new Stack<>();

        for (int i = m - 1; i >= 0; i--) {
            while (!st.isEmpty() && st.peek() <= nums2[i]) {
                st.pop();
            }

            if (st.isEmpty()) {
                hm.put(nums2[i], -1);
            } else {
                hm.put(nums2[i], st.peek());
            }

            st.push(nums2[i]);
        }

        for (int i = 0; i < n; i++) {
            arr[i] = hm.get(nums1[i]);
        }

        return arr;
    }
}

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        n = len(nums1)
        m = len(nums2)
        arr = [-1]*n
        hm = {}
        st = []

        for i in range(m-1,-1,-1):
            while st and st[-1] <= nums2[i]:
                st.pop()

            if not st:
                hm[nums2[i]] = -1

            else:
                hm[nums2[i]] = st[-1]
            st.append(nums2[i])
        
        for i in range(n):
            arr[i] = hm[nums1[i]]

        return arr
